File: piper_learnning/src/moveit_more_target_point.cpp
#include <ros/ros.h>
#include <moveit/move_group_interface/move_group_interface.h>
#include <geometry_msgs/Pose.h>
#include <moveit_msgs/RobotTrajectory.h>

int main(int argc, char** argv)
{
    ros::init(argc, argv, "moveit_relative_move_xyz");
    ros::NodeHandle nh;
    ros::AsyncSpinner spinner(1);
    spinner.start();

    // 初始化 MoveGroup 接口
    moveit::planning_interface::MoveGroupInterface move_group("arm");
    std::string end_effector_link = move_group.getEndEffectorLink();

    // 配置参数
    move_group.setGoalPositionTolerance(0.001);
    move_group.setGoalOrientationTolerance(0.01);
    move_group.setMaxAccelerationScalingFactor(0.2);
    move_group.setMaxVelocityScalingFactor(0.2);
    move_group.allowReplanning(true);
    // 控制机械臂先回到初始化位置
    move_group.setNamedTarget("home");
    move_group.move();
    sleep(1);
    // 获取当前末端执行器的位姿
    geometry_msgs::Pose current_pose = move_group.getCurrentPose(end_effector_link).pose;
    ROS_INFO_STREAM("Current pose: \n" << current_pose);

    // 构造路径点列表
    std::vector<geometry_msgs::Pose> waypoints;

    // Step 1: Z 轴上移 10cm
    geometry_msgs::Pose pose_1 = current_pose;
    pose_1.position.z += 0.10;
    waypoints.push_back(pose_1);

    // Step 2: Y 轴右移 10cm（在 pose_1 基础上）
    geometry_msgs::Pose pose_2 = pose_1;
    pose_2.position.x += 0.1;
    waypoints.push_back(pose_2);

    // Step 3: X 轴前移 10cm（在 pose_2 基础上）
    geometry_msgs::Pose pose_3 = pose_2;
    pose_3.position.y += 0.05;
    waypoints.push_back(pose_3);

    // 规划笛卡尔路径
    moveit_msgs::RobotTrajectory trajectory;
    const double eef_step = 0.005;      // 每步最大末端位移
    const double jump_threshold = 0.0;  // 关节跳跃阈值
    double fraction = move_group.computeCartesianPath(waypoints, eef_step, jump_threshold, trajectory);

    ROS_INFO("Cartesian path planning completed %.2f%% of the path", fraction * 100.0);

    // 执行路径
    if (fraction > 0.9) {
        moveit::planning_interface::MoveGroupInterface::Plan plan;
        plan.trajectory_ = trajectory;
        if (move_group.execute(plan)) {
            ROS_INFO("XYZ stepwise Cartesian path executed successfully.");
        } else {
            ROS_WARN("Execution failed.");
        }
    } else {
        ROS_WARN("Cartesian path planning incomplete. Only %.2f%% achieved.", fraction * 100.0);
    }

    spinner.stop();
    ros::shutdown();
    return 0;
}

File: piper_learnning/scripts/moveit_more_target_point.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import rospy
import moveit_commander
import geometry_msgs.msg
import sys
import copy

def main():
    # 初始化 MoveIt Commander 和 ROS 节点
    moveit_commander.roscpp_initialize(sys.argv)
    rospy.init_node("moveit_more_point", anonymous=True)

    # 创建 MoveGroup 接口对象
    move_group = moveit_commander.MoveGroupCommander("arm")

    # 获取末端执行器链接名称
    end_effector_link = move_group.get_end_effector_link()

    # 设置容差、重规划、速度、加速度参数
    move_group.allow_replanning(True)
    move_group.set_goal_position_tolerance(0.001)
    move_group.set_goal_orientation_tolerance(0.01)
    move_group.set_max_velocity_scaling_factor(0.2)
    move_group.set_max_acceleration_scaling_factor(0.2)

    # 回到初始化位置 "home"
    move_group.set_named_target("home")
    move_group.go(wait=True)
    rospy.sleep(1)

    # 设置起始点、过渡点、目标点
    start_pose = geometry_msgs.msg.Pose()
    start_pose.orientation.x = 1.0
    start_pose.orientation.y = -5.69991e-05
    start_pose.orientation.z = 4.14520e-05
    start_pose.orientation.w = 6.84446e-07
    start_pose.position.x = 0.183
    start_pose.position.y = -7.051e-6
    start_pose.position.z = 0.1542

    mid_pose = geometry_msgs.msg.Pose()
    mid_pose.orientation.x = 1.0
    mid_pose.orientation.y = -3.88606e-06
    mid_pose.orientation.z = -7.30472e-05
    mid_pose.orientation.w = 5.33069e-07
    mid_pose.position.x = 0.245193
    mid_pose.position.y = 1.09666e-6
    mid_pose.position.z = 0.212405

    goal_pose = geometry_msgs.msg.Pose()
    goal_pose.orientation.x = 1.0
    goal_pose.orientation.y = -2.70514e-05
    goal_pose.orientation.z = 3.82349e-05
    goal_pose.orientation.w = -1.8513e-05
    goal_pose.position.x = 0.245196
    goal_pose.position.y = -4.95272e-6
    goal_pose.position.z = 0.323646

    # 构造路点列表
    waypoints = [copy.deepcopy(start_pose), copy.deepcopy(mid_pose), copy.deepcopy(goal_pose)]

    # 降低速度以保证精度
    move_group.set_max_velocity_scaling_factor(0.1)
    move_group.set_max_acceleration_scaling_factor(0.1)

    # 计算笛卡尔轨迹
    eef_step = 0.01
    jump_threshold = 0.0
    (plan, fraction) = move_group.compute_cartesian_path(waypoints, eef_step, jump_threshold)

    # 执行规划好的路径
    move_group.execute(plan, wait=True)

    moveit_commander.roscpp_shutdown()

if __name__ == '__main__':
    main()
